File: web/src/bridge.ts
// Messages between the tldraw page and the Swift host.

import type { TLEditorSnapshot } from 'tldraw'

/** Goes up with any change to this contract; the host refuses a page built for another version. */
export const PROTOCOL = 14

export interface LoadPayload {
  /** Identifies the image's draft: its file path. Also the asset `src` the page resolves to a URL. */
  key: string
  mimeType: string
  pixelWidth: number
  pixelHeight: number
  /** Longest side, in pixels, of the preview this image's draft is rendered at. */
  previewMaxPixel: number
  /** The image's draft as the host stored it, or null for a fresh canvas. */
  snapshot: TLEditorSnapshot | null
  /**
   * Where the editor sits inside the page: the annotator's frame. The page is laid out at the room
   * the frame may grow within and the editor is this part of it, so the image opens fitted to the
   * frame. Absent for a `build`, which shows nothing.
   */
  frame?: PageRect
}

/** A rect in the page's coordinates: CSS points of the web view, x from its left and y from its top. */
export interface PageRect {
  x: number
  y: number
  width: number
  height: number
}

/**
 * The picture the page draws when a zoom comes to rest: where the editor sits inside the page,
 * which is the annotator's frame, and where the image is drawn, both in the page's points. The
 * image rect is the one the host's stand-in is drawing, so the two pictures are the same rect by
 * construction. The page places the editor, moves the camera, and answers once it has painted.
 */
export interface ViewRequest {
  frame: PageRect
  image: PageRect
}

/** What the page painted, once it has: the editor's size and where the image is, in the page's points. */
export interface ViewResult {
  width: number
  height: number
  image: PageRect
}

/**
 * One annotation an agent supplied with `add?marks=`. Every number is a fraction of the image:
 * `x` and `y` from its top-left corner, `w` and `h` of its size, `x2` and `y2` where an arrow
 * points. `build` turns these into ordinary shapes, which the user then edits like their own.
 *
 * On a text mark `w` is the box the words wrap in, and it is optional: without it the box is the
 * room between `x` and the right edge. `h` is the wrap's, never the mark's.
 */
export interface Mark {
  type: 'ellipse' | 'rectangle' | 'arrow' | 'text'
  x: number
  y: number
  w?: number
  h?: number
  x2?: number
  y2?: number
  text?: string
  /** A color id from config.ts; the first color when absent. */
  color?: string
}

/** A point zoom keeps in place: a fraction of the window, x from the left and y from the top. */
export interface ZoomAnchor { x: number; y: number }

export interface ToolInfo { id: string; label: string; key: string; symbol: string }
export interface ColorInfo { id: string; hex: string }

/** What `park` returns: the annotations to store (null when there are none) and a rendering when they changed. */
export interface ParkResult {
  snapshot: TLEditorSnapshot | null
  preview: string | null
}

export interface ExportItem { key: string; snapshot: TLEditorSnapshot }
/** What `export` returns: the renderings that succeeded, and the error that stopped the run, if any. */
export interface ExportResult {
  items: { key: string; png: string }[]
  error: string | null
}

type NativeMessage =
  /**
   * The editor is mounted. Carries the protocol version, what the host's toolbar should offer,
   * and every colour a mark may be drawn in.
   */
  | { type: 'ready'; protocol: number; tools: ToolInfo[]; markColors: ColorInfo[] }
  /** The active tool or the colour the next mark will be drawn in changed. */
  | { type: 'tool'; tool: string | null; color: string }
  /** The image from `load` is on the canvas. */
  | { type: 'loaded'; key: string }
  /** Finished. `png` is the rendering, or null when nothing was drawn: the host copies the original. */
  | { type: 'done'; png: string | null }
  | { type: 'cancel' }
  | { type: 'log'; message: string }
  /** The current image's annotations changed; null means they were all removed. Sent shortly after each change. */
  | { type: 'draft'; key: string; snapshot: TLEditorSnapshot | null }
  /**
   * Zoom is the host's: it resizes the window. `factor` multiplies the current size; null asks for
   * the fitted size. `at` is the cursor, whose point both sides keep in place; null (the keyboard)
   * means the window's middle.
   */
  | { type: 'zoom'; factor: number | null; at: ZoomAnchor | null }
  /**
   * A double-click with the select tool over the picture rather than over a mark: zoom in twofold
   * on the point it names, or back to the fitted size from anywhere above it. The host's own
   * two-finger double tap does the same.
   */
  | { type: 'smartZoom'; at: ZoomAnchor }

declare global {
  interface Window {
    shotnote?: {
      load(payload: LoadPayload): void
      /** The current draft for the host to store, with a preview when it changed. Does not clear the canvas. */
      park(): Promise<ParkResult>
      /** Clears the canvas. Call `park` first to keep the annotations. */
      reset(): void
      /**
       * An agent's marks as a draft, without the editor being shown: the shapes go on the canvas,
       * the snapshot and a rendering come back, and whatever was on the canvas is put back.
       */
      build(payload: LoadPayload, marks: Mark[]): Promise<ParkResult>
      /** Renders each item's draft to PNG. */
      export(items: ExportItem[]): Promise<ExportResult>
      /**
       * The current image's annotations alone, on a transparent canvas covering the image, as a
       * PNG data URL; null when nothing is drawn. `maxPixel` caps its longest side. The host lays
       * it over the screenshot while a zoom is moving.
       */
      overlay(maxPixel: number): Promise<string | null>
      setTool(id: string): void
      /**
       * Draws the exact picture the host's zoom stand-in is showing, and answers once it has been
       * painted: the host waits for that answer before it takes the stand-in away.
       */
      setView(view: ViewRequest): Promise<ViewResult | null>
      /** Exports the current image and replies with `done`. */
      finish(): void
    }
    webkit?: { messageHandlers?: { shotnote?: { postMessage(msg: NativeMessage): void } } }
  }
}

export function postToNative(msg: NativeMessage) {
  const handler = window.webkit?.messageHandlers?.shotnote
  if (handler) handler.postMessage(msg)
  else console.log('[shotnote → native]', msg)
}

File: Sources/Bridge.swift
import Foundation

// Mirror of web/src/bridge.ts. Change both files together; nothing else crosses the boundary.
// `protocolVersion` goes up with any change to either side; a page built for another version is
// refused at `ready`, so a stale web/dist is an error line instead of silent no-ops.
let bridgeProtocolVersion = 8

/// Sent to the page as `window.shotnote.load(payload)`. `key` identifies the image's draft.
struct LoadPayload: Encodable, Equatable {
    /// The file path. Also the asset `src` the page resolves to a served URL.
    let key: String
    let mimeType: String
    let pixelWidth: Int
    let pixelHeight: Int
    let viewWidth: Double
    let viewHeight: Double
}

/// The picture the page should draw when a zoom comes to rest: how far the image is magnified
/// inside the window (1 fits it), the middle of the visible part as a fraction of the image, and
/// the size the host has laid the window out at. The page waits for that size, applies the view,
/// and answers once it has painted it, which is when the stand-in may go.
struct ViewRequest: Encodable, Equatable {
    let ratio: Double
    let x: Double
    let y: Double
    let width: Double
    let height: Double
}

/// One annotation an agent supplied with `add?marks=`. Every number is a fraction of the image:
/// `x` and `y` from its top-left corner, `w` and `h` of its size, `x2` and `y2` where an arrow
/// points, so a mark does not depend on the screenshot's pixel size. `Commands.marks(from:)`
/// checks them; the page turns them into ordinary shapes the user then edits like their own.
struct Mark: Encodable, Equatable {
    enum Kind: String, Encodable, CaseIterable { case ellipse, rectangle, arrow, text }

    let type: Kind
    let x: Double
    let y: Double
    var w: Double?
    var h: Double?
    var x2: Double?
    var y2: Double?
    var text: String?
    /// A color id from web/src/config.ts; the page uses its first color when this is absent.
    var color: String?
}

struct ToolInfo: Identifiable, Equatable {
    let id: String
    let label: String
    let key: String
    let symbol: String   // SF Symbol
}

struct ColorInfo: Identifiable, Equatable {
    let id: String
    let hex: String
}

/// Every call the host makes into the page, rendered as the JavaScript that makes it.
enum PageAPI: Equatable {
    /// `snapshot` is the stored draft's JSON, or nil for a fresh canvas.
    case load(LoadPayload, snapshot: Data?)
    case park
    case reset
    /// An agent's marks as a draft, with nothing shown: the answer is a `ParkResult` to store.
    /// `snapshot` is the image's existing draft, which the marks are added to.
    case build(LoadPayload, snapshot: Data?, marks: [Mark])
    /// Each item's stored draft JSON, by key.
    case export([(key: String, snapshot: Data)])
    /// The current image's annotations alone, on a transparent canvas, no larger than this on the
    /// longest side: what the zoom stand-in lays over the screenshot.
    case overlay(maxPixel: Int)
    case setTool(String)
    case setColor(String)
    /// The picture to draw at the end of a zoom, and the answer that says it has been painted.
    case setView(ViewRequest)
    case finish

    /// `park`, `build`, `export`, and `setView` are async and return a value, so they run through
    /// `callAsyncJavaScript`; the rest are fire-and-forget. All guard on `window.shotnote` so a
    /// call that lands before the page's script runs is a no-op rather than an exception.
    var script: String {
        switch self {
        case .load(let payload, let snapshot):
            return "window.shotnote && window.shotnote.load(\(PageAPI.payload(payload, snapshot)));"
        case .park: return "return window.shotnote ? await window.shotnote.park() : null;"
        case .reset: return "window.shotnote && window.shotnote.reset();"
        case .build(let payload, let snapshot, let marks):
            return "return window.shotnote ? await window.shotnote.build(\(PageAPI.payload(payload, snapshot)),\(PageAPI.json(marks))) : null;"
        case .export(let items):
            let list = items.map { "{\"key\":\(PageAPI.json($0.key)),\"snapshot\":\(String(data: $0.snapshot, encoding: .utf8) ?? "null")}" }
            return "return window.shotnote ? await window.shotnote.export([\(list.joined(separator: ","))]) : null;"
        case .overlay(let maxPixel):
            return "return window.shotnote ? await window.shotnote.overlay(\(maxPixel)) : null;"
        case .setTool(let id): return "window.shotnote && window.shotnote.setTool(\(PageAPI.json(id)));"
        case .setColor(let id): return "window.shotnote && window.shotnote.setColor(\(PageAPI.json(id)));"
        case .setView(let view): return "return window.shotnote ? await window.shotnote.setView(\(PageAPI.json(view))) : null;"
        case .finish: return "window.shotnote && window.shotnote.finish();"
        }
    }

    static func == (a: PageAPI, b: PageAPI) -> Bool { a.script == b.script }

    /// The `load` object: the payload with the stored draft's JSON spliced in, since a snapshot is
    /// JSON the host never decodes. `build` takes the same object.
    static func payload(_ payload: LoadPayload, _ snapshot: Data?) -> String {
        let text = snapshot.flatMap { String(data: $0, encoding: .utf8) } ?? "null"
        return "{\"snapshot\":\(text),\(json(payload).dropFirst())"
    }

    /// JSON is valid JavaScript for objects, arrays, and strings; `withoutEscapingSlashes` keeps paths readable.
    static func json<T: Encodable>(_ value: T) -> String {
        let enc = JSONEncoder()
        enc.outputFormatting = [.withoutEscapingSlashes, .sortedKeys]
        guard let data = try? enc.encode(value), let text = String(data: data, encoding: .utf8) else { return "null" }
        return text
    }
}

/// Received from the page via `window.webkit.messageHandlers.shotnote.postMessage(...)`.
enum WebMessage {
    /// The editor is mounted. Carries the page's protocol version, what the toolbar should offer
    /// (`colors` is empty while the palette is hidden), and every color a pushed mark may name.
    case ready(protocol: Int, tools: [ToolInfo], colors: [ColorInfo], markColors: [ColorInfo])
    /// The active tool or color changed.
    case tool(tool: String?, color: String)
    /// The image from `load` is on the canvas.
    case loaded(key: String)
    /// Finished; a nil PNG means nothing was drawn and the original is what gets copied.
    case done(png: Data?)
    case cancel
    case log(String)
    /// The current image's annotations changed; a nil snapshot means they were all removed.
    case draft(key: String, snapshot: Any?)
    /// Multiply the window size by `factor`; nil asks for the fitted size. `at` is the cursor, a
    /// fraction of the window with y from the top, whose point zoom keeps in place; nil (the
    /// keyboard) means the window's middle.
    case zoom(factor: Double?, at: CGPoint?)

    init?(body: Any) {
        guard let dict = body as? [String: Any], let type = dict["type"] as? String else { return nil }
        switch type {
        case "ready":
            let tools = (dict["tools"] as? [[String: Any]] ?? []).compactMap { t -> ToolInfo? in
                guard let id = t["id"] as? String, let label = t["label"] as? String, let key = t["key"] as? String, let symbol = t["symbol"] as? String else { return nil }
                return ToolInfo(id: id, label: label, key: key, symbol: symbol)
            }
            func colors(_ key: String) -> [ColorInfo] {
                (dict[key] as? [[String: Any]] ?? []).compactMap { c -> ColorInfo? in
                    guard let id = c["id"] as? String, let hex = c["hex"] as? String else { return nil }
                    return ColorInfo(id: id, hex: hex)
                }
            }
            self = .ready(protocol: dict["protocol"] as? Int ?? 0, tools: tools, colors: colors("colors"), markColors: colors("markColors"))
        case "tool":
            self = .tool(tool: dict["tool"] as? String, color: dict["color"] as? String ?? "")
        case "loaded":
            guard let key = dict["key"] as? String else { return nil }
            self = .loaded(key: key)
        case "cancel": self = .cancel
        case "log": self = .log(dict["message"] as? String ?? "")
        case "done":
            guard let raw = dict["png"] else { return nil }
            if raw is NSNull { self = .done(png: nil) }
            else if let text = raw as? String, let data = Self.pngData(text) { self = .done(png: data) }
            else { return nil }
        case "draft":
            guard let key = dict["key"] as? String, dict.keys.contains("snapshot") else { return nil }
            let snapshot = dict["snapshot"]
            self = .draft(key: key, snapshot: snapshot is NSNull ? nil : snapshot)
        case "zoom":
            guard let raw = dict["factor"] else { return nil }
            let at = Self.point(dict["at"])
            if raw is NSNull { self = .zoom(factor: nil, at: at) }
            else if let n = raw as? NSNumber, n.doubleValue.isFinite, n.doubleValue > 0 { self = .zoom(factor: n.doubleValue, at: at) }
            else { return nil }
        default: return nil
        }
    }

    /// A unit point the page sent, or nil when it sent none: a keyboard step names no cursor, and
    /// a half-written one must not move the window on its own.
    static func point(_ raw: Any?) -> CGPoint? {
        guard let dict = raw as? [String: Any],
              let x = (dict["x"] as? NSNumber)?.doubleValue, x.isFinite,
              let y = (dict["y"] as? NSNumber)?.doubleValue, y.isFinite else { return nil }
        return CGPoint(x: x, y: y)
    }

    static func pngData(_ dataUrl: String) -> Data? {
        Data(base64Encoded: dataUrl.replacingOccurrences(of: "data:image/png;base64,", with: ""))
    }

    /// For the log when a body does not decode: its type and keys, never its content, which can be megabytes.
    static func describe(_ body: Any) -> String {
        guard let dict = body as? [String: Any] else { return "non-object \(type(of: body))" }
        return "type=\(dict["type"] as? String ?? "?") keys=\(dict.keys.sorted().joined(separator: ","))"
    }
}

/// What `PageAPI.park` returns: the draft to store (nil when the canvas has no annotations) and
/// a rendering when the user changed it since the host last saw one.
struct ParkResult {
    let snapshot: Any?
    let preview: Data?

    init?(body: Any?) {
        guard let dict = body as? [String: Any] else { return nil }
        let snapshot = dict["snapshot"]
        self.snapshot = snapshot is NSNull ? nil : snapshot
        self.preview = (dict["preview"] as? String).flatMap(WebMessage.pngData)
    }
}

/// What `PageAPI.export` returns: the renderings that succeeded, and the error that stopped the run.
struct ExportResult {
    let pngs: [String: Data]
    let error: String?

    init?(body: Any?) {
        guard let dict = body as? [String: Any] else { return nil }
        var pngs: [String: Data] = [:]
        for item in dict["items"] as? [[String: Any]] ?? [] {
            if let key = item["key"] as? String, let text = item["png"] as? String, let data = WebMessage.pngData(text) { pngs[key] = data }
        }
        self.pngs = pngs
        self.error = dict["error"] as? String
    }
}
